mypy/checker.py

```diff
@@ -4053,6 +4095,120 @@ def replay_lookup(new_parent_type: ProperType) -> Optional[Type]:
+        if singleton_index == -1:
+            singleton_index = possible_singleton_indices[-1]
+
+        enum_name = None
+        if isinstance(singleton, LiteralType) and singleton.is_enum_literal():
+            enum_name = singleton.fallback.type.fullname
+
+        target_type = [TypeRange(singleton, is_upper_bound=False)]
+
+        partial_type_maps = []
+        for i in chain_indices:
+            # If we try refining a singleton against itself, conditional_type_map
+            # will end up assuming that the 'else' branch is unreachable. This is
+            # typically not what we want: generally the user will intend for the
+            # singleton type to be some fixed 'sentinel' value and will want to refine
+            # the other exprs against this one instead.
+            if i == singleton_index:
+                continue
+
+            # Naturally, we can't refine operands which are not permitted to be refined.
+            if i not in narrowable_operand_indices:
+                continue
+
+            expr = operands[i]
+            expr_type = coerce_to_literal(operand_types[i])
+
+            if enum_name is not None:
+                expr_type = try_expanding_enum_to_union(expr_type, enum_name)
+            partial_type_maps.append(conditional_type_map(expr, expr_type, target_type))
+
+        return reduce_partial_type_maps(partial_type_maps)
+
+    def refine_equality_comparison_expression(self,
+                                              operands: List[Expression],
+                                              operand_types: List[Type],
+                                              chain_indices: List[int],
+                                              narrowable_operand_indices: Set[int],
+                                              ) -> Tuple[TypeMap, TypeMap]:
+        """Produces conditional type maps refining expressions used in an equality comparison.
+
+        For more details, see the docstring of 'refine_equality_comparison' up above.
+        The only difference is that this function is for refining equality operations
+        (e.g. 'a == b == c') instead of identity ('a is b is c').
+        """
+        non_optional_types = []
+        for i in chain_indices:
+            typ = operand_types[i]
+            if not is_optional(typ):
+                non_optional_types.append(typ)
+
+        # Make sure we have a mixture of optional and non-optional types.
+        if len(non_optional_types) == 0 or len(non_optional_types) == len(chain_indices):
+            return {}, {}
+
+        if_map = {}
+        for i in narrowable_operand_indices:
+            expr_type = operand_types[i]
+            if not is_optional(expr_type):
+                continue
+            if any(is_overlapping_erased_types(expr_type, t) for t in non_optional_types):
+                if_map[operands[i]] = remove_optional(expr_type)
+
+        return if_map, {}
```

Comment:
Should this refine both `x` and `y` if I have:
```python
x: Optional[int]
y: Optional[int]
if x == y == 1:
    ...
```
If yes, please add a test for this.

Reply:
It does! I added a test to `check-optional.test`.